Refactor CublasScopedContextHandler

keyradical opened the pull request:
There were a few different problems with the current way of things. **It's the best to look chronologically what each commit does.** I tried to follow the approach of "solve one problem and that will hopefully make it easier to fix later ones" which I think worked. 

The fixes: 

1. We don't need to include the UR header. We can replace the type of the `cublas_handle` map to be `CUcontext` instead of `ur_context_handle_t` ~ commit 1
2. Nowadays, we only use the primary context in sycl. So if we have only one device, the map will have only one entry: CuDevicePrimaryCtx -> CublasHandle. We also don't need to check if the current cuda context is primary or not. It's always primary. ~ commit 2
3. We were using `ContextSetExtendedDeleter` in the past I think for performance reasons. In the past, when we were also using other cuda contexts in sycl, the `cublas_handle` map would have more entries with more handles. We wanted for some of those entries to be deleted when the instance of `CublasScopedContextHandler` goes out of scope. This is not required anymore due to the use of primary context. ~ commit 3, 4
4. In our map, we don't need to have the `CublasHandle_t` to be wrapped into `std::atomic` and created on a heap with `new`. We actually can just have the map of `CUdevice -> CublasHandle_t` ~ commit 5, 6
5. To do the previous point we would need to modify the `cublas_handle` struct which we share with HIPSYCL. We could avoid that by dropping the use of it at all and use `static thread_local shared_ptr<unordered_map>` with custom deleter which destroys the cublas handles when the thread is destroyed ~ commit 7

Commit:
Replaced type of cublas_handle map to CUcontext to remove UR dependency

Commit:
Removed checking if current Ctx is not Primary. It's always primary Ctx

Commit:
Removed sycl::context* member and call to ContextSetExtendedDeleter. We don't need to do any early cleanup upon sycl context destruction.

Commit:
Remove unnecessary includes

Commit:
Changed handle_helper to have unordered_map of CUdevice(s) -> cublasHandle_t

Commit:
Switched to using shared_ptr instead of a cublas_handle struct with unordered map

hdelan commented on src/blas/backends/cublas/cublas_scope_handle_hipsycl.cpp:
```diff
@@ -71,4 +71,4 @@ CUstream CublasScopedContextHandler::get_stream(const sycl::queue& queue) {
 } // namespace cublas
 } // namespace blas
 } // namespace mkl
-} // namespace oneapi
```
Maybe remove this from your diff.

hdelan commented on src/blas/backends/cublas/cublas_task.hpp:
```diff
@@ -35,18 +35,6 @@
 #else
 #include "cublas_scope_handle_hipsycl.hpp"
 
-// After Plugin Interface removal in DPC++ ur.hpp is the new include
```
Great stuff

hdelan commented on src/blas/backends/cublas/cublas_scope_handle.hpp:
```diff
@@ -84,22 +63,15 @@ the handle must be destroyed when the context goes out of scope. This will bind
-    bool needToRecover_;
     sycl::interop_handle& ih;
-#ifdef ONEMKL_PI_INTERFACE_REMOVED
-    static thread_local cublas_handle<ur_context_handle_t> handle_helper;
```
It seems that you are removing all uses of `cublas_handle`, meaning the class is never instantiated. Is this intended?

keyradical replied:
I only did that in the last commit when I switched to using the `shared_ptr` instead of `cublas_handle`. After that change we don't need `cublas_handle` at all. But I'm not sure if we want to do that. 

The only problem with sticking to `cublas_handle` is that if I modify it's definition ([removing std::atomic commit](https://github.com/konradkusiak97/oneMKL/pull/1/commits/0eb8feaf0a794e82d079e2eb6e88ba9340808be2)) then we are breaking HIPSYCL implementation and would need to fix that.

And the base of all those changes is the fact that we only use cuda primary ctx in DPC++, I don't know what HIPSYCL does so maybe they need the `std::atomic` in the `cublas_handle`.

hdelan commented on src/blas/backends/cublas/cublas_scope_handle.cpp:
```diff
@@ -35,106 +30,46 @@ namespace cublas {
+        device_handle_map = std::shared_ptr<std::unordered_map<CUdevice, cublasHandle_t>>(
+            new std::unordered_map<CUdevice, cublasHandle_t>(), [](auto* map) {
+                for (auto& handle_pair : *map) {
+                    cublasStatus_t err;
```
Would you need to set a native context here?

keyradical replied:
I don't think so. Why do you think I would have to?

hdelan replied:
Is it valid to destroy the `cuBlasHandle_t` for device 1 when device 0 is the active device (the active native context)?

keyradical replied:
That is a great question

hdelan replied:
From https://docs.nvidia.com/cuda/cublas/index.html?highlight=cublasdestroy#cublasdestroy

> This function releases hardware resources used by the cuBLAS library. This function is usually the last call with a particular handle to the cuBLAS library. Because [cublasCreate()](https://docs.nvidia.com/cuda/cublas/index.html?highlight=cublasdestroy#cublascreate) allocates some internal resources and the release of those resources by calling [cublasDestroy()](https://docs.nvidia.com/cuda/cublas/index.html?highlight=cublasdestroy#cublasdestroy) will implicitly call cudaDeviceSynchronize(), it is recommended to minimize the number of times these functions are called.

This means we don't need to worry about cases where we might destroy the handle while work is still being executed on the device.

keyradical replied:
Yes, but also this will only be called at the end of the program anyway, I mean when the host thread is destroyed. So should be good. 

Regarding setting the native context, I think you're right. I need do need to set it here.

keyradical replied:
Now when I think about it, maybe this code is still too overcomplicated. Do we need the map at all? The current map is `thread_local` but also a single host thread is assigned to max. 1 cuda device. Or is it not?

If a single host thread can have multiple cublas handles tied to different cuda devices then fine, map is necessary. But cublas docs say:

>This approach allows the user to explicitly control the library setup when using multiple host threads and multiple GPUs. For example, the application can use cudaSetDevice() to associate different devices with different host threads and in each of those host threads it can initialize a unique handle to the cuBLAS library context, which will use the particular device associated with that host thread. Then, the cuBLAS library function calls made with different handles will automatically dispatch the computation to different devices.

Which suggests it's just 1 host thread -> 1 cuda device = 1 cublas handle per host thread.

hdelan reviewed (commented):
Great stuff. Main concern is whether `cublas_handle` is being used, and if it is needed at all.

Commit:
Remove map clear since the delete will handle that. And set primary Ctx tied to each device before destroying cublas handle

Commit:
Aborted changes in cublas_handle

cublas_scope_handle.cpp doesn't use cublas_handle anymore but the
hipsycl version does. So it's best to leave it as it is for now.

Commit:
Leave only 1 device and 1 cublasHandle_t as members of CublasScopedContextHandler

Since the mapping is now: thread_local map<device, cublasHandle> and we
have 1 host thread per 1 device, then the map would always have only 1
entry.

Commit:
Removed old comment
